### PR TITLE
Implementation of API Backend

### LINKED ISSUE #10: API backend: REST API
Implement restful api for `APIBackend` to store Trials in remote application.

Epic https://github.com/IceKhan13/purplecaffeine/issues/2

### PR DESCRIPTION
**Pull request by AnsahMohammad**
## Details and comments
Implemented simple API backend using Django Rest Framework.
can CREATE, READ, UPDATE, and DELETE Trial data.

- created Model to store Trial data
- implemented view.py to support CRUD operations of Trial

### Docker
- [x] Add a Docker env for development
- [x] Add a launching script
- [x] Update README

### CI
- [x] Fix tox
- [x] Update var FOLDER

---
Closes #10

### CODE REVIEW

**Comment by mickahell** [comment]
When you would make your checks passed, just read the comment in the log details. They are mainly about coding style and name. Use black gonna clear like half of the warning ;)

File: api_server/.gitignore
Comment (mickahell) [inline api_server/.gitignore:1]: You can add them into this file ;) https://github.com/IceKhan13/purplecaffeine/blob/main/.gitignore

File: api_server/PurpleCaffeine/settings.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+# Build paths inside the project like this: BASE_DIR / 'subdir'.
+BASE_DIR = Path(__file__).resolve().parent.parent
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = 'django-insecure-kb4_a&w(02n4h&5vcuiytouf6iub57$!s&l=uh5z_dlr7=*l67'
```
Comment (mickahell) [inline api_server/PurpleCaffeine/settings.py]: You'll have to make a variable of it

**Comment by mickahell** [comment]
Hi @AnsahMohammad :) my comments are more suggestions than review. I never used Django, I let the final word to Iskandar ;) except the few comments I already put could you :
- Add a small readme with the commands in order to launch the API (or add a Makefile [example](https://github.com/ESNFranceG33kTeam/CosmoAppy/blob/main/makefile)) and what do we need as a local db [example](https://github.com/ESNFranceG33kTeam/CosmoAppy#setup)
- For the text params of your file `settings.yml`, I thing it could be usefull to externalize them in a `.yaml` (something like that [1st example](https://github.com/ESNFranceG33kTeam/proto_mb/blob/main/conf/conf.yaml) or something like that [2nd example](https://github.com/ESNFranceG33kTeam/CosmoAppy/blob/main/test/conf_docker.yaml)). In this file you could have a default value for everything and then when we gonna deploy the app on the server we could put an extra file with the real value.
- Add a swagger fonction to generate the endpoint with the doc ([Django swagger doc](https://django-rest-swagger.readthedocs.io/en/latest/)) (can be in another issue)

Also I have a question :
- How do you create the db ? Can you add a fonction in order to automatically create the db if it doesn't exist (same with the tables) : something like `"CREATE DATABASE IF NOT EXISTS " + myconf.Namedb + ";"`and `"CREATE TABLE IF NOT EXISTS " + db_name + " (...);`

**Comment by IceKhan13** [comment]
@AnsahMohammad when you think this draft is ready for review convert it to PR and we will review it. Thanks!

File: api_server/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+COPY api_server/docker/start.sh /opt/
+
+RUN pip3 install --upgrade pip setuptools
+RUN pip3 install -r /opt/api_server/requirements.txt
+
+WORKDIR /opt/api_server
+
+EXPOSE 8000
+
+# I don't know how to launch Django without the webserver
```
Comment (IceKhan13) [inline api_server/docker/Dockerfile:44]: I'm usually going for `gunicorn` https://docs.djangoproject.com/en/4.2/howto/deployment/wsgi/gunicorn/

Basically run `gunicorn --bind 0.0.0.0:8000 myproject.wsgi`

File: api_server/purplecaffeine/settings.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = os.getenv("SERV_KEY")
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
```
Comment (IceKhan13) [inline api_server/purplecaffeine/settings.py]: Let's add here `int(os.environ.get("DEBUG", 0))`

File: api_server/purplecaffeine/settings.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+ALLOWED_HOSTS = []
```
Comment (IceKhan13) [inline api_server/purplecaffeine/settings.py]: Let's add here something like

```python
ALLOWED_HOSTS = os.environ.get("ALLOWED_HOSTS", "*").split(",")

```

**Review by IceKhan13 — CHANGES_REQUESTED** [review]
Looks very good! 🚀 

Couple of minor suggestions + style errors should be fixed :)

File: api_server/docker/Dockerfile (hunk 2 above)
Comment (mickahell) [reply api_server/docker/Dockerfile:44]: I just add this Dockerfile for preparing, but the Dockerize for production is in another issue #16

**Review by IceKhan13 — APPROVED** [review]
Let's go! Merging!

**Pull request merged**